### PR TITLE
feat: browser exporter

### PR DESCRIPTION
**Pull request by kermanx**
This PR introduces a new page on `/export` as an alternate to the `slidev export` CLI.

- A new "export" button is added under the "more options" menu in the navigation bar.
- When running `slidev`, the exporting URL will be shown.
- When running `slidev export`, a tip will be shown to inform the user to use the exporting UI.

> [!warning]
>
> Both PDF exporting and image capturing don't work well on Firefox. And I don't have a Safari to test.

### The Approach

- For PDF, use the browser's built-in "export to PDF" function.
- For PPTX/PNGs, the "Screen Capture API" is used.

### Screenshots (by 2024/12/5 11:55)

QQ_1733370981699

image

*On non-Chromium or chrome<94*:

image

### Todos

Maybe adding some UI for #1732?

### CODE REVIEW

**Review by antfu — APPROVED** [review]
Let's have it as experimental and ship it to iterates. I have some UI adjustment in mind but I think the feature wise is very solid already. Thanks a lot for working on it!

File: docs/custom/index.md [hunk 1]
```diff
@@ -27,6 +27,8 @@ keywords: keyword1,keyword2
 
 # enable presenter mode, can be boolean, 'dev' or 'build'
 presenter: true
+# enable exporting UI, can be boolean, 'dev' or 'build'
+exportingUI: dev
```
Comment (antfu) [inline docs/custom/index.md:31]: I think just `export` would be good enough :)

Comment (KermanX) [reply docs/custom/index.md:31]: We already have an `export` field in the headmatter:

```yaml
# export options
# use export CLI options in camelCase format
# Learn more: https://sli.dev/guide/exporting.html
export:
  format: pdf
  timeout: 30000
  dark: false
  withClicks: false
  withToc: false
```

Comment (antfu) [reply docs/custom/index.md:31]: How about `export.browser`?

Comment (antfu) [reply docs/custom/index.md:31]: I guess that would be fine. We could update the docs mentioning about that. The name `exportingUI` feels a bit verbose and we don't usually have the concept of `UI`. Or I don't know if we could call it `exporter` to differentiate with the CLI export

Comment (antfu) [reply docs/custom/index.md:31]: Or `browserExport` or `browserExporter` to be more specific?

**Comment by antfu** [comment]
Let's have it and iterate on the main. I am open for better naming

File: packages/slidev/node/cli.ts [hunk 2]
```diff
@@ -443,8 +443,19 @@ cli.command(
...
       const options = await resolveOptions({ entry: entryFile, theme }, 'export')
+
+      if (options.data.config.browserExporter !== false && !warned) {
+        warned = true
+        console.log(lightCyan('[Slidev] Try the new browser exporter!'))
+        console.log(
+          lightCyan('You can use the browser exporter instead by starting the dev server as normal and visit'),
+          `${blue('localhost:')}${dim('<port>')}${blue('/export')}\n`,
+        )
+      }
```
Comment (kalvdans) [inline packages/slidev/node/cli.ts:457]: This text is now spamming my terminal when I run in batch mode. Would you accept a PR that silences it in batch mode?

**Pull request merged**